agent/internal/agent.go

```diff
@@ -1,361 +1,240 @@
+		case errors.Is(err, context.Canceled):
+			return nil
+		case err != nil:
+			return err
 		default:
-			panic(fmt.Sprintf("unknown message received: %+v", msg))
-		}
-
-	case aproto.ContainerStateChanged:
-		if a.socket != nil {
-			ctx.Ask(a.socket, api.WriteMessage{Message: aproto.MasterMessage{ContainerStateChanged: &msg}})
-		} else {
-			ctx.Log().Warnf("Not sending container state change to the master: %+v", msg)
-		}
-	case aproto.ContainerLog:
-		if a.socket != nil {
-			ctx.Ask(a.socket, api.WriteMessage{Message: aproto.MasterMessage{ContainerLog: &msg}})
-		}
-
-	case aproto.ContainerStatsRecord:
-		if a.socket != nil {
-			ctx.Ask(a.socket, api.WriteMessage{Message: aproto.MasterMessage{ContainerStatsRecord: &msg}})
-		}
-
-	case model.TaskLog:
-		return a.postTaskLog(msg)
-
-	case actor.ChildFailed:
-		switch msg.Child {
-		case a.socket:
-			if a.attemptReconnect(ctx) {
-				return nil
-			}
-			ctx.Log().Warn("master socket disconnected, shutting down agent...")
-			a.onConnectionLost(ctx)
-		case a.cm:
-			ctx.Log().Warn("container manager failed, shutting down agent...")
-		case a.fluent:
-			ctx.Log().Warn("fluent bit failed, restarting it...")
-			go func() {
-				// Do this in a goroutine so we don't block the agent actor while retrying.
-				if err := a.restartFluent(ctx); err != nil {
-					logrus.WithError(err).Error("failed to restart fluent with retries")
-					ctx.Self().Stop()
-				}
-			}()
 			return nil
 		}
-		return errors.Wrapf(msg.Error, "unexpected child failure: %s", msg.Child.Address())
-
-	case actor.ChildStopped:
-		return errors.Errorf("unexpected child stopped: %s", msg.Child.Address())
+	})
 
-	case os.Signal:
-		switch msg {
-		case syscall.SIGINT, syscall.SIGTERM:
-			ctx.Log().Info("shutting down agent...")
-			ctx.Self().Stop()
-		default:
-			ctx.Log().Infof("unexpected signal received: %s", msg)
-		}
+	return a
+}
 
-	case echo.Context:
-		a.handleAPIRequest(ctx, msg)
+// Wait for the agent to exit, returning an error indicating the reason.
+func (a *Agent) Wait() error {
+	return a.wg.Wait()
+}
 
-	case actor.PostStop:
-		if a.fluent != nil {
-			if err := a.fluent.StopAndAwaitTermination(); err != nil {
-				ctx.Log().Errorf("error killing logging container %v", err)
-			}
+// Run sets up the agent and starts the watch loop. All configurations and system depenencies should
+// be setup _before_ the watch loop is started.
+func (a *Agent) run(ctx context.Context) error {
+	a.log.Trace("connecting to master")
+	socket, err := a.connect(ctx, false)
+	if err != nil {
+		return masterConnectionError{cause: fmt.Errorf("initial connection to master failed: %w", err)}
+	}
+	defer func() {
+		if socket == nil {
+			return
 		}
 
-		ctx.Log().Info("agent shut down")
-
-	default:
-		return actor.ErrUnexpectedMessage(ctx)
-	}
-	return nil
-}
+		a.log.Trace("cleaning up socket")
+		if cErr := socket.Close(); err != nil {
+			a.log.WithError(cErr).Error("closing master websocket")
```

Comment:
yeah, i usually try to do "failed to ..." or "crashed because..." at the top level (log isn't returned) and "closing ..." or "trying to ..." at intermediary levels. so good catch, i messed up my own pattern here.